src/asm/random.rs

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//
+// Copyright (c) 2022 Amazon.com, Inc. or its affiliates.
+//
+// Author(s):
+//   - Ali Saidi <alisaidi@amazon.com>
+
+#[derive(Copy, Clone, Debug)]
+
+/// Implement an interface for accessing Arm v8.5 RNG instructions
+/// A empty struct is used to confirm that the system has the
+/// instructions available.
+/// Example:
+/// ```no_run
+/// use cortex_a::asm::random::ArmRng;
+/// if let Some(rng) = ArmRng::new() {
+///    let rand_num = rng.rndr();
+/// }
+/// ```
+pub struct ArmRng(());
+
+use crate::registers::ID_AA64ISAR0_EL1;
+use core::arch::asm;
+use tock_registers::interfaces::Readable;
+
+impl ArmRng {
+    /// Return a empty object that is used to gate calling
+    /// rndr and rndrss on discovery of the feature so each
+    /// call doesn't need to confirm it
+    #[inline]
+    pub fn new() -> Option<Self> {
+        #[cfg(not(target_arch = "aarch64"))]
+        return None;
+
+        #[cfg(target_arch = "aarch64")]
+        if ID_AA64ISAR0_EL1.is_set(ID_AA64ISAR0_EL1::RNDR) {
+            Some(ArmRng(()))
+        } else {
+            None
+        }
+    }
+
+    /// Return an random number from the Arm v8.5 RNG
+    /// This returns an option because the instruction can fail
+    /// (e.g. the entropy is exhausted or the RNG has failed
+    #[inline]
+    pub fn rndr(self) -> Option<u64> {
```

Comment:
Should this be `&self`? I think it can make sense to keep an instance around for calling this function several times?

I guess with compiler optimizations and the empty struct, the end result is the same, but makes the intention clearer?